File: cs50/pset2/crack.c
#include <stdio.h>
#include <stdlib.h>
#include <cs50.h>
#include <string.h>
#include <crypt.h>

bool compare (string str, string hash);

int main(int argc, string argv[])
{
    if (argc != 2) // проверка на 2 агрумента командной строки
    {
        printf("Введите одно значение ключа\n"); // если аргументов не 2 - вернули ошибку
        return 1;
    }
    // подготовка к работе
    string hash = argv[1];
    char salt[3];
    salt[0] = hash[0];
    salt[1] = hash[1];
    salt[2] = '\0';
    char pass[6] = "\0\0\0\0\0\0";
    // string pass = crypt("Hello", salt);
    // printf("%c\n", salt[0]);
    // printf("%c\n", salt[1]);
    // printf("%s\n", pass);

// строка с буквами алфавита
    char letters[52];
    for (int i = 0; i < 26; i++)
    {
        letters[i] = (char) (65 + i);
        letters[i + 26] = (char) (97 + i);
    }
        // printf("%s\n", letters);

// проверка 1-буквенного пароля
    for (int i = 0; i < 52; i++)
    {
        pass[0] = letters[i];
        if ( compare(crypt(pass, hash), hash) )
        {
            printf("%s\n", pass);
            exit(0);
        }
    }
    printf("First step ended\n");

// проверка 2-буквенного пароля
    for (int i = 0; i < 52; i++)
    {
        pass[0] = letters[i];
        for (int j = 0; j < 52; j++)
        {
            pass[1] = letters[j];
            if ( compare(crypt(pass, hash), hash) )
            {
                printf("%s\n", pass);
                exit(0);
            }
        }
    }
    printf("Second step ended\n");

// проверка 3-буквенного пароля
    for (int i = 0; i < 52; i++)
    {
        pass[0] = letters[i];
        for (int j = 0; j < 52; j++)
        {
            pass[1] = letters[j];
            for (int k = 0; k < 52; k++)
            {
                pass[2] = letters[k];
                if ( compare(crypt(pass, hash), hash) )
                {
                    printf("%s\n", pass);
                    exit(0);
                }
            }
        }
    }
    printf("Third step ended\n");

// проверка 4-буквенного пароля
    for (int i = 0; i < 52; i++)
    {
        pass[0] = letters[i];
        for (int j = 0; j < 52; j++)
        {
            pass[1] = letters[j];
            for (int k = 0; k < 52; k++)
            {
                pass[2] = letters[k];
                for (int l = 0; l < 52; l++)
                {
                    pass[3] = letters[l];
                    if ( compare(crypt(pass, hash), hash) )
                    {
                        printf("%s\n", pass);
                        exit(0);
                    }
                }
            }
        }
    }
    printf("Fourth step ended\n");

// проверка 5-буквенного пароля
    for (int i = 0; i < 52; i++)
    {
        pass[0] = letters[i];
        for (int j = 0; j < 52; j++)
        {
            pass[1] = letters[j];
            for (int k = 0; k < 52; k++)
            {
                pass[2] = letters[k];
                for (int l = 0; l < 52; l++)
                {
                    pass[3] = letters[l];
                    for (int m = 0; m < 52; m ++)
                    {
                        pass[4] = letters[m];
                        if ( compare(crypt(pass, hash), hash) )
                        {
                            printf("%s\n", pass);
                            exit(0);
                        }
                    }
                }
            }
        }
    }
    printf("Fifth step ended, couldn't crack password\n");

    printf("%s\n", crypt("YES", "50"));


    // // int y = strcmp(str,st);
    // pass[0] = letters[19];
    // pass[1] = letters[5];
    // printf("%s\n", crypt(pass, hash));
    // if ( compare(crypt(pass, hash), hash) )
    //     {
    //         printf("%s\n", pass);
    //     }


}

bool compare (string str, string hash)
{
    int counter = 0;
    for (int i = 0; i < strlen(hash); i++)
    {
        if (str[i] == hash[i])
        {
            counter++;
        }
        else
        {
            break;
        }
    }
    if (counter == strlen(hash))
    {
        return true;
    }
    else
    {
        return false;
    }
}

File: cs50/pset6/crack.py
/****************************************************************************
 
* crack.py
*
 
* Provotorov Daniil
* provotorov.d.i@gmail.com
*
 
* Crack up to 5 signs password 
*
 
***************************************************************************/

# импортировали модуль crypt
import crypt
# импортировали аргументы командной строки
from sys import argv
# импортировали get_string
from cs50 import get_string
# проверили на число аргументов
if len(argv) != 2:
    print('Enter only 2 command-line arguments')
    exit(1)

hash = argv[1]
salt = hash[0] + hash[1]
all_characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
password = ""

# проверка 1-буквенного пароля
for i in range (len(all_characters)):
    password = all_characters[i]
    if crypt.crypt(password, salt) == hash:
        print(password)
        exit(0)
print("End 1-st step")

# проверка 2-буквенного пароля
for i in range (len(all_characters)):
    for j in range (len(all_characters)):
        password = all_characters[i] + all_characters[j]
        if crypt.crypt(password, salt) == hash:
            print(password)
            exit(0)
print("End 2-nd step")

# проверка 3-буквенного пароля
for i in range (len(all_characters)):
    for j in range (len(all_characters)):
        for k in range (len(all_characters)):
            password = all_characters[i] + all_characters[j] + all_characters[k]
            if crypt.crypt(password, salt) == hash:
                print(password)
                exit(0)
print("End 3-rd step")

# проверка 4-буквенного пароля
for i in range (len(all_characters)):
    for j in range (len(all_characters)):
        for k in range (len(all_characters)):
            for l in range (len(all_characters)):
                password = all_characters[i] + all_characters[j] + all_characters[k] + all_characters[l]
                if crypt.crypt(password, salt) == hash:
                    print(password)
                    exit(0)
print("End 4-th step")

# проверка 5-буквенного пароля
for i in range (len(all_characters)):
    for j in range (len(all_characters)):
        for k in range (len(all_characters)):
            for l in range (len(all_characters)):
                for m in range (len(all_characters)):
                    password = all_characters[i] + all_characters[j] + all_characters[k] + all_characters[l] + all_characters[m]
                    if crypt.crypt(password, salt) == hash:
                        print(password)
                        exit(0)
print("End 5-th step")

print("Couldn't crack password")